File: Projecto/proto/Pathfinder.py
#coding: utf-8

import gclasses

class PathFinder:
	up = gclasses.Vect(0, -1)
	down = gclasses.Vect(0, 1)
	right = gclasses.Vect(1, 0)
	left = gclasses.Vect(-1, 0)

	def __init__(self, parent, depart, arrivee):
		self.grille = parent
		self.actif_node = None
		self.depart = depart
		self.arrivee = arrivee
		self.nodelist = []

	def run(self, pos = None):
		found = False
		stop = False
		true_pos = pos
		if pos == None:
			true_pos = self.depart
			dnode = Node(self.grille, self, true_pos)
			self.nodelist.append(dnode)

		localnodelist = []
		
		#print("position:", self.possibilities(true_pos))
		for sens in self.possibilities(true_pos):
			print("vers: ", sens)
			node = Node(self.grille, self, true_pos + sens)
			print(node.getVect())
			for n in self.nodelist:
				print(n.getVect())
			if node in self.nodelist:
				print(node.getVect(), "already in nodelist")
				pass
			else:
				self.nodelist.append(node)
				localnodelist.append(node)
				if node.hcost == 0:
					found = True
					break

		if len(localnodelist) == 0:
			stop = True

		if not stop:
			if not found:
				if len(localnodelist) != 0:
					for n in localnodelist:
						return self.run(n.pos)
				else:
					return None
			else:
				return self.nodelist

	#def get(self):
	#	return self.nodelist

	def possibilities(self, pos):
		possibilities_list = []

		#merdier pour vérifier si on peut aller en haut, avec tous les cas possibles
		if ((pos + PathFinder.up)[0] > 0 and (pos + PathFinder.up)[0] < 14 and (pos + PathFinder.up)[1] > 0 and (pos + PathFinder.up)[1] < 14):
			if self.grille.getValue(pos, PathFinder.up) != "w" and self.grille.getValue(pos, PathFinder.up) != "0":
				if ((pos + PathFinder.up*2)[0] > 0 and (pos + PathFinder.up*2)[0] < 14 and (pos + PathFinder.up*2)[1] > 0 and (pos + PathFinder.up*2)[1] < 14):
					if self.grille.getValue(pos, PathFinder.up*2) == "p":
						if ((pos + PathFinder.up*3)[0] > 0 and (pos + PathFinder.up*3)[0] < 14 and (pos + PathFinder.up*3)[1] > 0 and (pos + PathFinder.up*3)[1] < 14):
							if self.grille.getValue(pos, PathFinder.up*3) != "w" and self.grille.getValue(pos, PathFinder.up*3) != "0":
								if ((pos + PathFinder.up*4)[0] > 0 and (pos + PathFinder.up*4)[0] < 14 and (pos + PathFinder.up*4)[1] > 0 and (pos + PathFinder.up*4)[1] < 14):
									possibilities_list.append(PathFinder.up*4)
							else:
								if ((pos + PathFinder.up*2+PathFinder.left)[0] > 0 and (pos + PathFinder.up*2+PathFinder.left)[0] < 14 and (pos + PathFinder.up*2+PathFinder.left)[1] > 0 and (pos + PathFinder.up*2+PathFinder.left)[1] < 14):
									if self.grille.getValue(pos, PathFinder.up*2+PathFinder.left) != "w" and self.grille.getValue(pos, PathFinder.up*2+PathFinder.left) != "0":
										if ((pos + PathFinder.up*2+PathFinder.left*2)[0] > 0 and (pos + PathFinder.up*2+PathFinder.left*2)[0] < 14 and (pos + PathFinder.up*2+PathFinder.left*2)[1] > 0 and (pos + PathFinder.up*2+PathFinder.left*2)[1] < 14):
											possibilities_list.append(PathFinder.up*2+PathFinder.left*2)
								if ((pos + PathFinder.up*2+PathFinder.right)[0] > 0 and (pos + PathFinder.up*2+PathFinder.right)[0] < 14 and (pos + PathFinder.up*4+PathFinder.right)[1] > 0 and (pos + PathFinder.up*2+PathFinder.right)[1] < 14):
									if self.grille.getValue(pos, PathFinder.up*2+PathFinder.right) != "w" and self.grille.getValue(pos, PathFinder.up*2+PathFinder.right) != "0":
										if ((pos + PathFinder.up*2+PathFinder.right*2)[0] > 0 and (pos + PathFinder.up*2+PathFinder.right*2)[0] < 14 and (pos + PathFinder.up*2+PathFinder.right*2)[1] > 0 and (pos + PathFinder.up*2+PathFinder.right*2)[1] < 14):
											possibilities_list.append(PathFinder.up*2+PathFinder.right*2)
					else:
						possibilities_list.append(PathFinder.up*2)

		#pareil pour down
		if ((pos + PathFinder.down)[0] > 0 and (pos + PathFinder.down)[0] < 14 and (pos + PathFinder.down)[1] > 0 and (pos + PathFinder.down)[1] < 14):
			if self.grille.getValue(pos, PathFinder.down) != "w" and self.grille.getValue(pos, PathFinder.down) != "0":
				if ((pos + PathFinder.down*2)[0] > 0 and (pos + PathFinder.down*2)[0] < 14 and (pos + PathFinder.down*2)[1] > 0 and (pos + PathFinder.down*2)[1] < 14):
					if self.grille.getValue(pos, PathFinder.down*2) == "p":
						if ((pos + PathFinder.down*3)[0] > 0 and (pos + PathFinder.down*3)[0] < 14 and (pos + PathFinder.down*3)[1] > 0 and (pos + PathFinder.down*3)[1] < 14):
							if self.grille.getValue(pos, PathFinder.down*3) != "w":
								if ((pos + PathFinder.down*4)[0] > 0 and (pos + PathFinder.down*4)[0] < 14 and (pos + PathFinder.down*4)[1] > 0 and (pos + PathFinder.down*4)[1] < 14):
									possibilities_list.append(PathFinder.down*4)
							else:
								if ((pos + PathFinder.down*2+PathFinder.left)[0] > 0 and (pos + PathFinder.down*2+PathFinder.left)[0] < 14 and (pos + PathFinder.down*2+PathFinder.left)[1] > 0 and (pos + PathFinder.down*2+PathFinder.left)[1] < 14):
									if self.grille.getValue(pos, PathFinder.down*2+PathFinder.left) != "w" and self.grille.getValue(pos, PathFinder.down*2+PathFinder.left) != "0":
										if ((pos + PathFinder.down*2+PathFinder.left*2)[0] > 0 and (pos + PathFinder.down*2+PathFinder.left*2)[0] < 14 and (pos + PathFinder.down*2+PathFinder.left*2)[1] > 0 and (pos + PathFinder.down*2+PathFinder.left*2)[1] < 14):
											possibilities_list.append(PathFinder.down*2+PathFinder.left*2)
								if ((pos + PathFinder.down*2+PathFinder.right)[0] > 0 and (pos + PathFinder.down*2+PathFinder.right)[0] < 14 and (pos + PathFinder.down*2+PathFinder.right)[1] > 0 and (pos + PathFinder.down*2+PathFinder.right)[1] < 14):
									if self.grille.getValue(pos, PathFinder.down*2+PathFinder.right) != "w" and self.grille.getValue(pos, PathFinder.down*2+PathFinder.right) != "0":
										if ((pos + PathFinder.down*2+PathFinder.right*2)[0] > 0 and (pos + PathFinder.down*2+PathFinder.right*2)[0] < 14 and (pos + PathFinder.down*2+PathFinder.right*2)[1] > 0 and (pos + PathFinder.down*2+PathFinder.right*2)[1] < 14):
											possibilities_list.append(PathFinder.down*2+PathFinder.right*2)
					else:
						possibilities_list.append(PathFinder.down*2)

		#pareil pour right
		if ((pos + PathFinder.right)[0] > 0 and (pos + PathFinder.right)[0] < 14 and (pos + PathFinder.right)[1] > 0 and (pos + PathFinder.right)[1] < 14):
			if self.grille.getValue(pos, PathFinder.right) != "w" and self.grille.getValue(pos, PathFinder.right) != "0":
				if ((pos + PathFinder.right*2)[0] > 0 and (pos + PathFinder.right*2)[0] < 14 and (pos + PathFinder.right*2)[1] > 0 and (pos + PathFinder.right*2)[1] < 14):
					if self.grille.getValue(pos, PathFinder.right*2) == "p":
						if ((pos + PathFinder.right*3)[0] > 0 and (pos + PathFinder.right*3)[0] < 14 and (pos + PathFinder.right*3)[1] > 0 and (pos + PathFinder.right*3)[1] < 14):
							if self.grille.getValue(pos, PathFinder.right*3) != "w":
								if ((pos + PathFinder.right*4)[0] > 0 and (pos + PathFinder.right*4)[0] < 14 and (pos + PathFinder.right*4)[1] > 0 and (pos + PathFinder.right*4)[1] < 14):
									possibilities_list.append(PathFinder.right*4)
							else:
								if ((pos + PathFinder.right*2+PathFinder.up)[0] > 0 and (pos + PathFinder.right*2+PathFinder.up)[0] < 14 and (pos + PathFinder.right*2+PathFinder.up)[1] > 0 and (pos + PathFinder.right*2+PathFinder.up)[1] < 14):
									if self.grille.getValue(pos, PathFinder.right*2+PathFinder.up) != "w" and self.grille.getValue(pos, PathFinder.right*2+PathFinder.up) != "0":
										if ((pos + PathFinder.right*2+PathFinder.up*2)[0] > 0 and (pos + PathFinder.right*2+PathFinder.up*2)[0] < 14 and (pos + PathFinder.right*2+PathFinder.up*2)[1] > 0 and (pos + PathFinder.right*2+PathFinder.up*2)[1] < 14):
											possibilities_list.append(PathFinder.right*2+PathFinder.up*2)
								if ((pos + PathFinder.right*2+PathFinder.down)[0] > 0 and (pos + PathFinder.right*4+PathFinder.down)[0] < 14 and (pos + PathFinder.right*2+PathFinder.down)[1] > 0 and (pos + PathFinder.right*2+PathFinder.down)[1] < 14):
									if self.grille.getValue(pos, PathFinder.right*4+PathFinder.down) != "w" and self.grille.getValue(pos, PathFinder.right*2+PathFinder.down) != "0":
										if ((pos + PathFinder.right*2+PathFinder.down*2)[0] > 0 and (pos + PathFinder.right*2+PathFinder.down*2)[0] < 14 and (pos + PathFinder.right*2+PathFinder.down*2)[1] > 0 and (pos + PathFinder.right*2+PathFinder.down*2)[1] < 14):
											possibilities_list.append(PathFinder.right*2+PathFinder.down*2)
					else:
						possibilities_list.append(PathFinder.right*2)

		#pareil pour left
		if ((pos + PathFinder.left)[0] > -1 and (pos + PathFinder.left)[0] < 14 and (pos + PathFinder.left)[1] > -1 and (pos + PathFinder.left)[1] < 14):
			if self.grille.getValue(pos, PathFinder.left) != "w" and self.grille.getValue(pos, PathFinder.left) != "0":
				if ((pos + PathFinder.left*2)[0] > -1 and (pos + PathFinder.left*2)[0] < 14 and (pos + PathFinder.left*2)[1] > -1 and (pos + PathFinder.left*2)[1] < 14):
					if self.grille.getValue(pos, PathFinder.left*2) == "p":
						if ((pos + PathFinder.left*3)[0] > -1 and (pos + PathFinder.left*3)[0] < 14 and (pos + PathFinder.left*3)[1] > -1 and (pos + PathFinder.left*3)[1] < 14):
							if self.grille.getValue(pos, PathFinder.left*3) != "w":
								if ((pos + PathFinder.left*4)[0] > -1 and (pos + PathFinder.left*4)[0] < 14 and (pos + PathFinder.left*4)[1] > -1 and (pos + PathFinder.left*4)[1] < 14):
									possibilities_list.append(PathFinder.left*4)
							else:
								if ((pos + PathFinder.left*2+PathFinder.up)[0] > -1 and (pos + PathFinder.left*2+PathFinder.up)[0] < 14 and (pos + PathFinder.left*2+PathFinder.up)[1] > -1 and (pos + PathFinder.left*2+PathFinder.up)[1] < 14):
									if self.grille.getValue(pos, PathFinder.left*2+PathFinder.up) != "w" and self.grille.getValue(pos, PathFinder.left*2+PathFinder.up) != "0":
										if ((pos + PathFinder.left*2+PathFinder.up*2)[0] > -1 and (pos + PathFinder.left*2+PathFinder.up*2)[0] < 14 and (pos + PathFinder.left*2+PathFinder.up*2)[1] > -1 and (pos + PathFinder.left*2+PathFinder.up*2)[1] < 14):
											possibilities_list.append(PathFinder.left*2+PathFinder.up*2)
								if ((pos + PathFinder.left*2+PathFinder.down)[0] > -1 and (pos + PathFinder.left*2+PathFinder.down)[0] < 14 and (pos + PathFinder.left*2+PathFinder.down)[1] > -1 and (pos + PathFinder.left*2+PathFinder.down)[1] < 14):
									if self.grille.getValue(pos, PathFinder.left*2+PathFinder.down) != "w" and self.grille.getValue(pos, PathFinder.left*2+PathFinder.down) != "0":
										if ((pos + PathFinder.left*2+PathFinder.down*2)[0] > -1 and (pos + PathFinder.left*2+PathFinder.down*2)[0] < 14 and (pos + PathFinder.left*2+PathFinder.down*2)[1] > -1 and (pos + PathFinder.left*2+PathFinder.down*2)[1] < 14):
											possibilities_list.append(PathFinder.left*2+PathFinder.down*2)
					else:
						possibilities_list.append(PathFinder.left*2)

		return possibilities_list

class Node:
	def __init__(self, grille, parent, position, prec = None):
		self.parent = prec
		self.grille = grille
		self.pos = position
		self.gcost = gclasses.Math.distance(self.pos, parent.depart)
		self.hcost = gclasses.Math.distance(self.pos, parent.arrivee)
		self.cost = self.gcost + self.hcost

	def __eq__(self, node):
		if self.pos == node.pos:
			return True
		else:
			return False

	def __get__(self):
		print(self.cost)

	def __str__(self):
		return str(self.cost)

	def getTree(self):
		tree = [self]
		if self.parent:
			for n in self.parent.getTree():
				tree.append(n)
		return tree

	def getVect(self):
		return self.pos

File: TestsU/TestWall/quoridor/PathFinder.java
package quoridor;

public class PathFinder {
	private Vector up = new Vector(0, -1);
	public Vector down = new Vector(0, 1);
	private final Vector right = new Vector(1, 0);
	private final Vector left = new Vector(-1, 0);
	private int x = 1;

	public Board grille;
	public PFNode actif_node;
	public Vector depart;
	public Vector arrivee;
	public PFNode [] nodelist;
	
	

	public PathFinder(Board parent, Vector depart, Vector arrivee){
		grille = parent;
		this.depart = depart;
		this.arrivee = arrivee;
		nodelist = new PFNode[x];
	}

	public PFNode[] run(Vector pos){
		boolean found = false;
		boolean stop = false;
		Vector true_pos = null;
		PFNode dep_node;
		if (pos != null){
			true_pos = depart;
			dep_node = new PFNode(grille,this ,null);
			nodelist[0] = dep_node;
		}

		PFNode [] local_node_list = new PFNode[x];

		Vector [] possibilities = possibilities(true_pos);
		for (int i = 0; i < possibilities.length; i++){
			Vector sens = possibilities[i];

			PFNode node = new PFNode(grille, this, true_pos.add(sens));

			boolean exist = false;
			for (int j = 0; j < nodelist.length; j++){
				if (node.get() == nodelist[j].get())
					exist = true;
			}

			if (exist == false){
				nodelist[i] = node;
				local_node_list[i] = node;
				if (node.hcost == 0){
					found = true;
					break;
				}
			}
		}

		if (local_node_list.length == 0)
			stop = true;

		if (stop == false){
			if (found == false){
				if (local_node_list.length != 0){
					for (int i = 0; i < local_node_list.length; i++){
						PFNode [] l = run(local_node_list[i].pos);

						//rallonger la liste
						PFNode[] copy = nodelist;
						x++;
						nodelist = new PFNode[x];
						for (int j = 0; j < copy.length; j++)
							nodelist[j] = copy[j];

						if (l[0] != null){
							return l;
						}
					}
				}
				else
					return null;
			}
			else
				return nodelist;
		}
		return null;
	}
	
	

	/*
	public static Node[] get(){
		return nodelist;
	}
	*/
		public Vector [] possibilities(Vector pos) {
		int a = 1;
		Vector [] possibilities_list = new Vector[a];

		if (pos.add(up).y > -1 && pos.add(up).y < 17  && pos.add(up).x > -1 && pos.add(up).x < 17){
			if (grille.getValue(pos, up) != "w"){
				if (pos.add(up.mul(2)).y > -1 && pos.add(up.mul(3)).y < 17 && pos.add(up.mul(3)).x > -1 && pos.add(up.mul(3)).y < 17){
					if (grille.getValue(pos, up.mul(3)) != "w"){
						if (pos.add(up.mul(2)).y > -1 && pos.add(up.mul(2)).y < 17 && pos.add(up.mul(2)).x > -1 && pos.add(up.mul(2)).x < 17){
							if (grille.getValue(pos, up.mul(2)) == "p"){
								if (pos.add(up.mul(3)).y > -1 && pos.add(up.mul(3)).y < 17 && pos.add(up.mul(3)).x > -1 && pos.add(up.mul(3)).x < 17){
									if (grille.getValue(pos, up.mul(3)) != "w"){
										if (pos.add(up.mul(4)).y > -1 && pos.add(up.mul(4)).y < 17 && pos.add(up.mul(4)).x > -1 && pos.add(up.mul(4)).y < 17){
											possibilities_list[a-1] = up.mul(4);

											a++;
											Vector[] copy = possibilities_list;
											possibilities_list = new Vector[a];
											for (int i = 0; i < a-2; i++)
												possibilities_list[i] = copy[i];
										}
									}
									else{
										if (pos.add(up.mul(2)).add(left).y > -1 && pos.add(up.mul(2)).add(left).y < 17 && pos.add(up.mul(2)).add(left).x > -1 && pos.add(up.mul(2)).add(left).x <17 ){
											if (grille.getValue(pos, up.mul(2).add(left)) != "w"){
												if (pos.add(up.mul(2)).add(left.mul(2)).y > -1 && pos.add(up.mul(2)).add(left.mul(2)).y <17  && pos.add(up.mul(2)).add(left.mul(2)).x > -1 && pos.add(up.mul(2)).add(left.mul(2)).x <17){
													possibilities_list[a-1] = up.mul(4);

													a++;
													Vector[] copy = possibilities_list;
													possibilities_list = new Vector[a];
													for (int i = 0; i < a-2; i++)
														possibilities_list[i] = copy[i];
												}
											}
										}
										if (pos.add(up.mul(2)).add(right).y > -1 && pos.add(up.mul(2)).add(right).y <17  && pos.add(up.mul(2)).add(right).x > -1 && pos.add(up.mul(2)).add(right).x <17 ){
											if (grille.getValue(pos, up.mul(2).add(right)) != "w"){
												if (pos.add(up.mul(2)).add(right.mul(2)).y > -1 && pos.add(up.mul(2)).add(right.mul(2)).y <17  && pos.add(up.mul(2)).add(right.mul(2)).x > -1 && pos.add(up.mul(2)).add(right.mul(2)).x <17){
													possibilities_list[a-1] = up.mul(4);

													a++;
													Vector[] copy = possibilities_list;
													possibilities_list = new Vector[a];
													for (int i = 0; i < a-2; i++)
														possibilities_list[i] = copy[i];
												}
											}
										}
									}
								}
							}
							else{
								possibilities_list[a-1] = up.mul(2);

								a++;
								Vector[] copy = possibilities_list;
								possibilities_list = new Vector[a];
								for (int i = 0; i < a-2; i++)
									possibilities_list[i] = copy[i];
							}
						}
					}
				}
			}
		}

		if (pos.add(down).y > -1 && pos.add(down).y < 17 && pos.add(down).x > -1 && pos.add(down).x <17 ){
			if (grille.getValue(pos, down) != "w"){
				if (pos.add(down.mul(2)).y > -1 && pos.add(down.mul(3)).y < 17 && pos.add(down.mul(3)).x > -1 && pos.add(down.mul(3)).y <17 ){
					if (grille.getValue(pos, down.mul(3)) != "w"){
						if (pos.add(down.mul(2)).y > -1 && pos.add(down.mul(2)).y < 17 && pos.add(down.mul(2)).x > -1 && pos.add(down.mul(2)).x < 17){
							if (grille.getValue(pos, down.mul(2)) == "p"){
								if (pos.add(down.mul(3)).y > -1 && pos.add(down.mul(3)).y < 17 && pos.add(down.mul(3)).x > -1 && pos.add(down.mul(3)).x <17 ){
									if (grille.getValue(pos, down.mul(3)) != "w"){
										if (pos.add(down.mul(4)).y > -1 && pos.add(down.mul(4)).y < 17 && pos.add(down.mul(4)).x > -1 && pos.add(down.mul(4)).y <17 ){
											possibilities_list[a-1] = down.mul(4);

											a++;
											Vector[] copy = possibilities_list;
											possibilities_list = new Vector[a];
											for (int i = 0; i < a-2; i++)
												possibilities_list[i] = copy[i];
										}
									}
									else{
										if (pos.add(down.mul(2)).add(left).y > -1 && pos.add(down.mul(2)).add(left).y < 17 && pos.add(down.mul(2)).add(left).x > -1 && pos.add(down.mul(2)).add(left).x <17 ){
											if (grille.getValue(pos, down.mul(2).add(left)) != "w"){
												if (pos.add(down.mul(2)).add(left.mul(2)).y > -1 && pos.add(down.mul(2)).add(left.mul(2)).y <17  && pos.add(down.mul(2)).add(left.mul(2)).x > -1 && pos.add(down.mul(2)).add(left.mul(2)).x <17){
													possibilities_list[a-1] = down.mul(4);

													a++;
													Vector[] copy = possibilities_list;
													possibilities_list = new Vector[a];
													for (int i = 0; i < a-2; i++)
														possibilities_list[i] = copy[i];
												}
											}
										}
										if (pos.add(down.mul(2)).add(right).y > -1 && pos.add(down.mul(2)).add(right).y < 17 && pos.add(down.mul(2)).add(right).x > -1 && pos.add(down.mul(2)).add(right).x < 17){
											if (grille.getValue(pos, down.mul(2).add(right)) != "w"){
												if (pos.add(down.mul(2)).add(right.mul(2)).y > -1 && pos.add(down.mul(2)).add(right.mul(2)).y <17  && pos.add(down.mul(2)).add(right.mul(2)).x > -1 && pos.add(down.mul(2)).add(right.mul(2)).x <17){
													possibilities_list[a-1] = down.mul(4);

													a++;
													Vector[] copy = possibilities_list;
													possibilities_list = new Vector[a];
													for (int i = 0; i < a-2; i++)
														possibilities_list[i] = copy[i];
												}
											}
										}
									}
								}
							}
							else{
								possibilities_list[a-1] = down.mul(2);

								a++;
								Vector[] copy = possibilities_list;
								possibilities_list = new Vector[a];
								for (int i = 0; i < a-2; i++)
									possibilities_list[i] = copy[i];
							}
						}
					}
				}
			}
		}

		if (pos.add(right).y > -1 && pos.add(right).y <17  && pos.add(right).x > -1 && pos.add(right).x <17 ){
			if (grille.getValue(pos, right) != "w"){
				if (pos.add(right.mul(2)).y > -1 && pos.add(right.mul(3)).y <17  && pos.add(right.mul(3)).x > -1 && pos.add(right.mul(3)).y <17 ){
					if (grille.getValue(pos, right.mul(3)) != "w"){
						if (pos.add(right.mul(2)).y > -1 && pos.add(right.mul(2)).y < 17 && pos.add(right.mul(2)).x > -1 && pos.add(right.mul(2)).x < 17){
							if (grille.getValue(pos, right.mul(2)) == "p"){
								if (pos.add(right.mul(3)).y > -1 && pos.add(right.mul(3)).y < 17 && pos.add(right.mul(3)).x > -1 && pos.add(right.mul(3)).x < 17){
									if (grille.getValue(pos, right.mul(3)) != "w"){
										if (pos.add(right.mul(4)).y > -1 && pos.add(right.mul(4)).y < 17 && pos.add(right.mul(4)).x > -1 && pos.add(right.mul(4)).y <17){
											possibilities_list[a-1] = right.mul(4);

											a++;
											Vector[] copy = possibilities_list;
											possibilities_list = new Vector[a];
											for (int i = 0; i < a-2; i++)
												possibilities_list[i] = copy[i];
										}
									}
									else{
										if (pos.add(right.mul(2)).add(up).y > -1 && pos.add(right.mul(2)).add(up).y <17  && pos.add(right.mul(2)).add(up).x > -1 && pos.add(right.mul(2)).add(up).x <17 ){
											if (grille.getValue(pos, right.mul(2).add(up)) != "w"){
												if (pos.add(right.mul(2)).add(up.mul(2)).y > -1 && pos.add(right.mul(2)).add(up.mul(2)).y <  17&& pos.add(right.mul(2)).add(up.mul(2)).x > -1 && pos.add(right.mul(2)).add(up.mul(2)).x <17){
													possibilities_list[a-1] = right.mul(4);

													a++;
													Vector[] copy = possibilities_list;
													possibilities_list = new Vector[a];
													for (int i = 0; i < a-2; i++)
														possibilities_list[i] = copy[i];
												}
											}
										}
										if (pos.add(right.mul(2)).add(down).y > -1 && pos.add(right.mul(2)).add(down).y <17  && pos.add(right.mul(2)).add(down).x > -1 && pos.add(right.mul(2)).add(down).x < 17){
											if (grille.getValue(pos, right.mul(2).add(down)) != "w"){
												if (pos.add(right.mul(2)).add(down.mul(2)).y > -1 && pos.add(right.mul(2)).add(down.mul(2)).y < 17 && pos.add(right.mul(2)).add(down.mul(2)).x > -1 && pos.add(right.mul(2)).add(down.mul(2)).x <177){
													possibilities_list[a-1] = right.mul(4);

													a++;
													Vector[] copy = possibilities_list;
													possibilities_list = new Vector[a];
													for (int i = 0; i < a-2; i++)
														possibilities_list[i] = copy[i];
												}
											}
										}
									}
								}
							}
							else{
								possibilities_list[a-1] = right.mul(2);

								a++;
								Vector[] copy = possibilities_list;
								possibilities_list = new Vector[a];
								for (int i = 0; i < a-2; i++)
									possibilities_list[i] = copy[i];
							}
						}
					}
				}
			}
		}

		if (pos.add(left).y > -1 && pos.add(left).y < 17 && pos.add(left).x > -1 && pos.add(left).x < 17){
			if (grille.getValue(pos, left) != "w"){
				if (pos.add(left.mul(2)).y > -1 && pos.add(left.mul(3)).y < 17 && pos.add(left.mul(3)).x > -1 && pos.add(left.mul(3)).y <17 ){
					if (grille.getValue(pos, left.mul(3)) != "w"){
						if (pos.add(left.mul(2)).y > -1 && pos.add(left.mul(2)).y < 17 && pos.add(left.mul(2)).x > -1 && pos.add(left.mul(2)).x <17 ){
							if (grille.getValue(pos, left.mul(2)) == "p"){
								if (pos.add(left.mul(3)).y > -1 && pos.add(left.mul(3)).y < 17 && pos.add(left.mul(3)).x > -1 && pos.add(left.mul(3)).x <17 ){
									if (grille.getValue(pos, left.mul(3)) != "w"){
										if (pos.add(left.mul(4)).y > -1 && pos.add(left.mul(4)).y < 17 && pos.add(left.mul(4)).x > -1 && pos.add(left.mul(4)).y < 17){
											possibilities_list[a-1] = left.mul(4);

											a++;
											Vector[] copy = possibilities_list;
											possibilities_list = new Vector[a];
											for (int i = 0; i < a-2; i++)
												possibilities_list[i] = copy[i];
										}
									}
									else{
										if (pos.add(left.mul(2)).add(right).y > -1 && pos.add(left.mul(2)).add(right).y <17  && pos.add(left.mul(2)).add(right).x > -1 && pos.add(left.mul(2)).add(right).x < 17){
											if (grille.getValue(pos, left.mul(2).add(right)) != "w"){
												if (pos.add(left.mul(2)).add(right.mul(2)).y > -1 && pos.add(left.mul(2)).add(right.mul(2)).y < 17 && pos.add(left.mul(2)).add(right.mul(2)).x > -1 && pos.add(left.mul(2)).add(right.mul(2)).x <17){
													possibilities_list[a-1] = left.mul(4);

													a++;
													Vector[] copy = possibilities_list;
													possibilities_list = new Vector[a];
													for (int i = 0; i < a-2; i++)
														possibilities_list[i] = copy[i];
												}
											}
										}
										if (pos.add(left.mul(2)).add(down).y > -1 && pos.add(left.mul(2)).add(down).y < 17 && pos.add(left.mul(2)).add(down).x > -1 && pos.add(left.mul(2)).add(down).x < 17){
											if (grille.getValue(pos, left.mul(2).add(down)) != "w"){
												if (pos.add(left.mul(2)).add(down.mul(2)).y > -1 && pos.add(left.mul(2)).add(down.mul(2)).y <17 && pos.add(left.mul(2)).add(down.mul(2)).x > -1 && pos.add(left.mul(2)).add(down.mul(2)).x <17){
													possibilities_list[a-1] = left.mul(4);

													a++;
													Vector[] copy = possibilities_list;
													possibilities_list = new Vector[a];
													for (int i = 0; i < a-2; i++)
														possibilities_list[i] = copy[i];
												}
											}
										}
									}
								}
							}
							else{
								possibilities_list[a-1] = left.mul(2);

								a++;
								Vector[] copy = possibilities_list;
								possibilities_list = new Vector[a];
								for (int i = 0; i < a-2; i++)
									possibilities_list[i] = copy[i];
							}
						}
					}
				}
			}
		}

		return possibilities_list;
	}
}